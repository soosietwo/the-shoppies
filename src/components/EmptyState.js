import React from "react";
import { EmptyState } from "@shopify/polaris";

const EmptyStateMessage = (movies) => {
  return !movies.length ? (
    <EmptyState
      heading="No movies found"
      action=""
      image="https://cdn.shopify.com/s/files/1/2376/3301/products/emptystate-files.png"
    >
      <p>Search to get started!</p>
    </EmptyState>
  ) : undefined;
};

export default EmptyStateMessage;
