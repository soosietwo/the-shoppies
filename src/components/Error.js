import React from "react";
import { Banner } from "@shopify/polaris";

// TODO: make this more user friendly and less scary
const Error = (props) => (
  <Banner status="critical" title={props.title}>
    <p>{props.details}</p>
  </Banner>
);

export default Error;
