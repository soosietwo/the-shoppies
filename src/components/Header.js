import React from "react";
import { TopBar } from "@shopify/polaris";
import SearchBar from "./SearchBar";

import { gql, useQuery } from "@apollo/client";

export const NOMINEES_CONNECTION_QUERY = gql`
  query NOMINEES_CONNECTION_QUERY {
    nomineesConnection {
      aggregate {
        count
      }
    }
  }
`;

const Header = ({ toggleSheetActive, nomineesCount }) => {
  const { loading, error, data } = useQuery(NOMINEES_CONNECTION_QUERY);
  const count = data ? data.nomineesConnection.aggregate.count : 0;

  return (
    <TopBar
      showNavigationToggle
      searchField={<SearchBar />}
      secondaryMenu={
        <TopBar.Menu
          activatorContent={
            <span>
              Nominees
              <span
                style={{
                  padding: "5px 9px",
                  borderRadius: "50%",
                  background: "red",
                  color: "white",
                  marginLeft: "5px",
                  fontSize: "1.2rem",
                }}
              >
                {count}
              </span>
            </span>
          }
          onOpen={toggleSheetActive}
          onClose={toggleSheetActive}
        />
      }
    />
  );
};

export default Header;
