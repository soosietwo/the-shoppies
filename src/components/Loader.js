import React from "react";
import {
  ResourceItem,
  TextContainer,
  SkeletonDisplayText,
  SkeletonBodyText,
  ResourceList,
  SkeletonThumbnail,
  Stack,
} from "@shopify/polaris";

const Loader = ({ count }) => {
  return (
    <ResourceList
      items={[...Array(count).keys()]}
      renderItem={() => (
        <ResourceItem>
          <Stack>
            <SkeletonThumbnail size="large" />
            <Stack.Item fill>
              <TextContainer>
                <SkeletonDisplayText size="small" />
                <SkeletonBodyText />
              </TextContainer>
            </Stack.Item>
          </Stack>
        </ResourceItem>
      )}
    />
  );
};

export default Loader;
