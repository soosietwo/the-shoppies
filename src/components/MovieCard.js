import React from "react";
import {
  Heading,
  Subheading,
  Button,
  ResourceItem,
  TextContainer,
} from "@shopify/polaris";
import { gql, useMutation } from "@apollo/client";

import { NOMINEES_QUERY } from "./Sidebar";

const ADD_NOMINEE_MUTATION = gql`
  mutation ADD_NOMINEE_MUTATION(
    $id: ID!
    $title: String!
    $poster: String!
    $year: Int!
  ) {
    addNominee(id: $id, title: $title, poster: $poster, year: $year) {
      id
      title
      poster
      year
    }
  }
`;

const MovieCard = (props) => {
  const {
    disableAction,
    actionText,
    action,
    movie: { title, year, poster, id },
  } = props;

  const posterMarkup = (
    <img
      style={{ maxWidth: "100px" }}
      src={poster}
      alt={`Poster for ${title}`}
    />
  );

  const [
    addNominee,
    { loading, data, error },
  ] = useMutation(ADD_NOMINEE_MUTATION, {
    refetchQueries: [{ query: NOMINEES_QUERY }],
  });

  return (
    <ResourceItem
      id={id}
      media={posterMarkup}
      accessibilityLabel={`View details for ${title}`}
    >
      <TextContainer>
        <Heading>{title}</Heading>
        <Subheading>{year}</Subheading>

        {action && (
          <Button
            disabled={disableAction && disableAction(props.movie)}
            onClick={() => {
              const variables = {
                id,
                poster,
                year,
                title,
              };

              addNominee({ variables });
            }}
          >
            {actionText}
          </Button>
        )}
      </TextContainer>
    </ResourceItem>
  );
};

export default MovieCard;
