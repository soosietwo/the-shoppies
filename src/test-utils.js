// test-utils.js
import React from "react";
import { render } from "@testing-library/react";
import { AppProvider, Frame } from "@shopify/polaris";
import en from "@shopify/polaris/locales/en.json";
import Store from "./store";

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

const AllTheProviders = ({ children }) => {
  return (
    <AppProvider i18n={en}>
      <Store>
        <Frame>{children}</Frame>
      </Store>
    </AppProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
