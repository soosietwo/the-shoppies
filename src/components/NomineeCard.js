import React from "react";
import {
  Heading,
  Subheading,
  Button,
  ResourceItem,
  TextContainer,
} from "@shopify/polaris";
import { gql, useMutation } from "@apollo/client";

import { NOMINEES_QUERY } from "./Sidebar";
import { NOMINEES_CONNECTION_QUERY } from "./Header";

const DELETE_NOMINEE_MUTATION = gql`
  mutation DELETE_NOMINEE_MUTATION($id: ID!) {
    deleteNominee(id: $id) {
      id
      title
      poster
      year
    }
  }
`;

const MovieCard = (props) => {
  const {
    movie: { title, year, poster, id },
  } = props;

  const posterMarkup = (
    <img
      style={{ maxWidth: "100px" }}
      src={poster}
      alt={`Poster for ${title}`}
    />
  );

  const [deleteNominee, { loading, data, error }] = useMutation(
    DELETE_NOMINEE_MUTATION,
    {
      refetchQueries: [
        { query: NOMINEES_QUERY },
        { query: NOMINEES_CONNECTION_QUERY },
      ],
    }
  );

  return (
    <ResourceItem
      id={id}
      media={posterMarkup}
      accessibilityLabel={`View details for ${title}`}
    >
      <TextContainer>
        <Heading>{title}</Heading>
        <Subheading>{year}</Subheading>
        <Button
          disabled={loading}
          onClick={() => deleteNominee({ variables: { id } })}
        >
          {loading ? "Removing..." : "Remove"}
        </Button>
      </TextContainer>
    </ResourceItem>
  );
};

export default MovieCard;
