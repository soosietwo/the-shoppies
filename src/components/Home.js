import React, { useContext, useCallback, useState } from "react";
import { Page, Card, Frame, Banner } from "@shopify/polaris";
import { gql, useQuery } from "@apollo/client";

import MoviesList from "./MoviesList";
import Pagination from "./Pagination";
import Loader from "./Loader";
import Error from "./Error";
import Sidebar from "./Sidebar";
import Header from "./Header";
import MovieCard from "./MovieCard";
import { Context } from "../store";

export const NOMINEES_QUERY = gql`
  query NOMINEES_QUERY {
    nominees {
      id
      title
      poster
      year
    }
  }
`;

const Home = () => {
  const [state] = useContext(Context);
  const [sheetActive, setSheetActive] = useState(false);
  const toggleSheetActive = useCallback(
    () => setSheetActive((sheetActive) => !sheetActive),
    []
  );

  const { loading, error, data } = useQuery(NOMINEES_QUERY);

  return (
    <Frame topBar={<Header toggleSheetActive={toggleSheetActive} />}>
      <Page>
        <Card>
          {state.error && !state.isLoading && (
            <Error title="Oops! Something went wrong" details={state.error} />
          )}

          {data && data.nominees.length === 5 && (
            <Banner
              status="success"
              title="All done!"
              action={{
                content: "Go to nominees",
                onAction: () => toggleSheetActive(true),
              }}
            >
              <p>You have chosen 5 nominees. Submit or remove them.</p>
            </Banner>
          )}

          {state.isLoading || loading ? (
            <Loader count={10} />
          ) : (
            <MoviesList
              singular="movie"
              plural="movies"
              movies={state.movies}
              totalResults={state.totalResults}
              emptyStateTitle="No movies found"
              emptyStateCaption="Start typing in your favourite movie titles to get started!"
              renderItem={(movie) => (
                <MovieCard
                  movie={movie}
                  nominees={(data && data.nominees) || []}
                  toggleSheetActive={toggleSheetActive}
                />
              )}
            />
          )}

          {state.totalResults > 10 && <Pagination />}
        </Card>

        <Sidebar
          nominees={(data && data.nominees) || []}
          toggleSheetActive={toggleSheetActive}
          sheetActive={sheetActive}
        />
      </Page>
    </Frame>
  );
};

export default Home;
