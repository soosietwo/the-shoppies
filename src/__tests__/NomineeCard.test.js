import React from "react";
import { render, fireEvent, waitFor } from "../test-utils";
import { MockedProvider } from "@apollo/client/testing";
import { ApolloConsumer } from "@apollo/client";
import NomineeCard, {
  DELETE_NOMINEE_MUTATION,
} from "../components/NomineeCard";
import { NOMINEES_QUERY } from "../components/Home";
import { NOMINEES_CONNECTION_QUERY } from "../components/Header";

const mockNominee = {
  id: "132",
  title: "test",
  poster: "123",
  year: 1234,
};

const mockMutationResult = jest.fn(() => ({
  data: {
    deleteNominee: {
      ...mockNominee,
    },
  },
}));

const mocks = [
  {
    request: { query: NOMINEES_QUERY },
    result: {
      data: {
        nominees: [mockNominee],
      },
    },
  },
  {
    request: { query: NOMINEES_CONNECTION_QUERY },
    result: {
      data: {
        nomineesConnection: {
          aggregate: {
            count: 1,
          },
        },
      },
    },
  },
  {
    request: {
      query: DELETE_NOMINEE_MUTATION,
      variables: { id: "132" },
    },
    result: mockMutationResult,
  },
  {
    request: { query: NOMINEES_QUERY },
    result: {
      data: {
        nominees: [],
      },
    },
  },
  {
    request: { query: NOMINEES_CONNECTION_QUERY },
    result: {
      data: {
        nomineesConnection: {
          aggregate: {
            count: 0,
          },
        },
      },
    },
  },
];

describe("NomineeCard", () => {
  it("renders and matches snapshot", () => {
    const { container } = render(
      <MockedProvider>
        <NomineeCard movie={mockNominee} nominees={[]} />
      </MockedProvider>
    );

    expect(container.querySelector(".Polaris-ResourceItem")).toMatchSnapshot();
  });

  it("can remove movie from nominees", async () => {
    let apolloClient;

    const { getByText } = render(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {(client) => {
            apolloClient = client;
            return <NomineeCard movie={mockNominee} />;
          }}
        </ApolloConsumer>
      </MockedProvider>
    );
    await apolloClient.query({ query: NOMINEES_QUERY });
    await apolloClient.query({ query: NOMINEES_CONNECTION_QUERY });

    fireEvent.click(getByText("Remove"));

    await waitFor(() => {
      expect(mockMutationResult).toHaveBeenCalled();
    });
  });
});
