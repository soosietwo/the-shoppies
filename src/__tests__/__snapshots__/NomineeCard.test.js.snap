// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NomineeCard renders and matches snapshot 1`] = `
<div
  class="Polaris-ResourceItem"
>
  <button
    aria-label="View details for test"
    class="Polaris-ResourceItem__Button"
    tabindex="0"
  />
  <div
    class="Polaris-ResourceItem__Container Polaris-ResourceItem--alignmentCenter"
    id="132"
  >
    <div
      class="Polaris-ResourceItem__Owned"
    >
      <div
        class="Polaris-ResourceItem__Media"
      >
        <img
          alt="Poster for test"
          src="123"
          style="max-width: 100px;"
        />
      </div>
    </div>
    <div
      class="Polaris-ResourceItem__Content"
    >
      <div
        class="Polaris-TextContainer"
      >
        <h2
          class="Polaris-Heading"
        >
          test
        </h2>
        <h3
          class="Polaris-Subheading"
        >
          1234
        </h3>
        <button
          class="Polaris-Button Polaris-Button--destructive"
          type="button"
        >
          <span
            class="Polaris-Button__Content"
          >
            <span
              class="Polaris-Button__Text"
            >
              Remove
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
